keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes operator, I would like to be able to declaratively control upgrades in a systematic fashion.
+
+#### Story 2
+
+As a Kubernetes operator, I would like to be able to declaratively control certificate renewal in a systematic fashion.
+
+#### Story 3
+
+As a Kubernetes operator, I would like to be able to declaratively control changes of the current cluster's settings in a systematic fashion.
+
+#### Story 4
+
+As a Kubernetes operator, I would like to be able to declaratively certificate authority rotation.
+
+#### Story 5 
+As a Kubernetes operator, I would like to control whether nodes are cordoned and drained for tasks or if they are performed "hot"
+
+### Implementation Details/Notes/Constraints
+
+The initial proposal for implementing the kubeadm operator can be summarized by the following sequence diagram; the goal of the current iteration is to validate/improve the proposed approach.
+
+![](20190916-kubeadm-operator.png)
+
+#### Action
+The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Action
+metadata:
+  name: action-sample
+spec:
+  upgrade: 
+      targetVersion: v1.16.0
+```
+
+or 
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: Action
+metadata:
+  name: action-sample
+spec:
+  renewCertificates: {}
+```
+#### Action Controller and TaskGroup
+The `Action controller` is a component of the kubeadm operator that manages the  `Action` CRD and implements all the domain-specific knowledge about the kubeadm workflows e.g. it knows that the certificate-renewal workflow requires to run `kubeadm alpha certs renew` on all the control plane nodes in the cluster.
+
+As a consequence, the  `Action controller`  generates one or more `TaskGroup` objects, each describing a `TaskTemplate` object - e.g. run `kubeadm alpha certs renew` -  targeting a set of nodes matching a given `NodeSelector` - e.g. all nodes with `node-role.kubernetes.io/master` label -.
+
+```yaml
+apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
+kind: TaskGroup
+metadata:
+  name: taskgroup-sample2
+  ownerReference:
+      ...
+spec:
+  nodeSelector: 
+    matchLabels:
+      node-role.kubernetes.io/master: ""
+  selector:
+    matchLabels:
+      app: a
+  template:
+    metadata:
+      labels:
+        app: a
+    alphaCertsRenew: {}
+```
+
+It is important to notice that the  `Action controller` is going to always generate `TaskGroup` in sequential order - only after the current `TaskGroup` is completed successfully a new one will be generated.
```

Comment:
What happens if the current `TaskGroup` does not succeed? Assuming an upgrade, if I understood correctly:

1. TaskGroup created for control plane nodes (`node-role.kubernetes.io/master: ""`)
2. When 1. has succeeded, TaskGroup created for worker nodes (unclear how they'll be targeted if not explicitly labeled -- with a `matchExpression` maybe?)

If 1. fails, will it be retried?; do you foresee a retrial count limit?. Will the `TaskGroup` be eventually removed?

Reply:
> What happens if the current TaskGroup does not succeed?

The current working assumption is to block the execution immediately after a task fail, and to not take initiative automatically. I'm working on a possible UX to let the user to choose if to retry or to skip the failed task.

> unclear how they'll be targeted if not explicitly labeled -- with a matchExpression maybe

matchExpression is an option, even if not currently supported by controller-tools. 
I donìt have a veto also on requiring additional labels for the operator (might be only during alpha)

Reply:
Still, we can try to do away with the Task objects (possibly even with Task Groups) and leverage DaemonSets and Jobs where part of these problems have been solved and kubectl is your friend.

Reply:
I have described pros and cons about using bare DaemonSets and Jobs the alternatives paragraphs, and TBH I don't that going on this path is a good idea, because we will be stuck as soon as we have a small requirement that deviates from DaemonSets and Jobs abstractions. And as far as I am concerned, we are already in this case for operability requirements (e.g. dry-run, pause, error management)

Reply:
[1]